File: containers/splunk/splunk__connectors/csv/python/deploy_csv_connectors.py
import os
import yaml
import subprocess
import argparse
import shutil
from datetime import datetime
from pydantic import BaseModel, validator
import psutil

class CSVConnectorsConfig(BaseModel):
    log_dir: str
    schemas_file: str

class SplunkConfig(BaseModel):
    enabled: bool
    hec_url: str
    hec_token: str

class ElasticsearchConfig(BaseModel):
    enabled: bool
    host: str
    index: str

class KubernetesConfig(BaseModel):
    replicas: int
    storage_class: str
    log_storage_size: str

class DeploymentConfig(BaseModel):
    method: str
    namespace: str
    kubernetes: KubernetesConfig

class Config(BaseModel):
    csv_connectors: CSVConnectorsConfig
    splunk: SplunkConfig
    elasticsearch: ElasticsearchConfig
    deployment: DeploymentConfig
    buffer_timeout: float
    worker_count: int
    batch_size: int
    delimiter: str

    @validator('worker_count')
    def validate_worker_count(cls, v):
        cpu_count = os.cpu_count() or 4
        if v > cpu_count * 2:
            raise ValueError(f"worker_count ({v}) exceeds twice the CPU count ({cpu_count})")
        return v

    @validator('deployment')
    def validate_deployment_method(cls, v):
        if v.method not in ['docker', 'podman', 'kubernetes', 'ansible']:
            raise ValueError(f"Invalid deployment method: {v.method}")
        return v

def validate_resources():
    cpu_count = os.cpu_count() or 4
    mem_total = psutil.virtual_memory().total / (1024 ** 3)  # GB
    disk_free = psutil.disk_usage('/').free / (1024 ** 3)  # GB
    if cpu_count < 2:
        print("Warning: Less than 2 CPU cores available")
    if mem_total < 2:
        print("Warning: Less than 2GB memory available")
    if disk_free < 2:
        print("Warning: Less than 2GB disk space available")

def backup_files(config_files, backup_dir):
    os.makedirs(backup_dir, exist_ok=True)
    for src in config_files:
        if os.path.exists(src):
            dst = os.path.join(backup_dir, os.path.basename(src))
            shutil.copy2(src, dst)

def deploy_docker(config, use_podman=False):
    compose_cmd = 'podman-compose' if use_podman else 'docker-compose'
    if shutil.which(compose_cmd) is None:
        raise RuntimeError(f"{compose_cmd} not found")

    env = os.environ.copy()
    env.update({
        'CSV_LOG_DIR': config.csv_connectors.log_dir,
        'SCHEMAS_FILE': config.csv_connectors.schemas_file,
        'SPLUNK_ENABLED': str(config.splunk.enabled).lower(),
        'SPLUNK_HEC_URL': config.splunk.hec_url,
        'SPLUNK_TOKEN': config.splunk.hec_token,
        'ES_ENABLED': str(config.elasticsearch.enabled).lower(),
        'ES_HOST': config.elasticsearch.host,
        'ES_INDEX': config.elasticsearch.index,
        'BATCH_SIZE': str(config.batch_size),
        'BUFFER_TIMEOUT': str(config.buffer_timeout),
        'WORKER_COUNT': str(config.worker_count),
        'CSV_DELIMITER': config.delimiter
    })

    subprocess.run([compose_cmd, 'up', '-d', '--build'], env=env, check=True)

def deploy_kubernetes(config):
    if shutil.which('kubectl') is None:
        raise RuntimeError("kubectl not found")

    manifests = ['csv-pvc.yaml', 'csv-splunk-deployment.yaml', 'csv-elasticsearch-deployment.yaml']
    for manifest in manifests:
        with open(manifest, 'r') as f:
            content = f.read()
            content = content.replace('{{namespace}}', config.deployment.namespace)
            content = content.replace('{{replicas}}', str(config.deployment.kubernetes.replicas))
            content = content.replace('{{storage_class}}', config.deployment.kubernetes.storage_class)
            content = content.replace('{{log_storage_size}}', config.deployment.kubernetes.log_storage_size)
            content = content.replace('{{splunk_hec_url}}', config.splunk.hec_url)
            content = content.replace('{{splunk_token}}', config.splunk.hec_token)
            content = content.replace('{{es_host}}', config.elasticsearch.host)
            content = content.replace('{{es_index}}', config.elasticsearch.index)
            content = content.replace('{{batch_size}}', str(config.batch_size))
            content = content.replace('{{buffer_timeout}}', str(config.buffer_timeout))
            content = content.replace('{{worker_count}}', str(config.worker_count))
            content = content.replace('{{delimiter}}', config.delimiter)
            with open(f"temp_{manifest}", 'w') as f_temp:
                f_temp.write(content)

    for manifest in manifests:
        subprocess.run(['kubectl', 'apply', '-f', f"temp_{manifest}"], check=True)
        os.remove(f"temp_{manifest}")

def deploy_ansible(config):
    if shutil.which('ansible-playbook') is None:
        raise RuntimeError("ansible-playbook not found")

    subprocess.run(['ansible-playbook', 'deploy_csv_connectors.yml', '-e', f"config_file={os.path.abspath('deploy_config.yaml')}"], check=True)

def cleanup_docker(use_podman=False):
    compose_cmd = 'podman-compose' if use_podman else 'docker-compose'
    if shutil.which(compose_cmd):
        subprocess.run([compose_cmd, 'down', '-v'], check=True)

def cleanup_kubernetes(config):
    if shutil.which('kubectl'):
        manifests = ['csv-pvc.yaml', 'csv-splunk-deployment.yaml', 'csv-elasticsearch-deployment.yaml']
        for manifest in manifests:
            subprocess.run(['kubectl', 'delete', '-f', manifest, '-n', config.deployment.namespace], check=False)

def main():
    parser = argparse.ArgumentParser(description='Deploy CSV connectors for Splunk and Elasticsearch')
    parser.add_argument('--config', default='deploy_config.yaml', help='Path to configuration file')
    parser.add_argument('--cleanup', action='store_true', help='Cleanup deployment')
    args = parser.parse_args()

    config_files = [
        'deploy_config.yaml',
        'docker-compose.yml',
        'Dockerfile.connector',
        'schemas.yaml',
        'csv_splunk_connector.py',
        'csv_elasticsearch_connector.py',
        'csv-splunk-deployment.yaml',
        'csv-elasticsearch-deployment.yaml',
        'csv-pvc.yaml',
        'deploy_csv_connectors.yml'
    ]

    if args.cleanup:
        print("Cleaning up CSV connectors deployment...")
        config = load_config(args.config)
        if config.deployment.method == 'docker':
            cleanup_docker()
        elif config.deployment.method == 'podman':
            cleanup_docker(use_podman=True)
        elif config.deployment.method == 'kubernetes':
            cleanup_kubernetes(config)
        elif config.deployment.method == 'ansible':
            deploy_ansible(config)
        return

    validate_resources()
    config = load_config(args.config)
    backup_dir = f"backup/{datetime.now().strftime('%Y%m%d_%H%M%S')}"
    backup_files(config_files, backup_dir)

    if config.deployment.method == 'docker':
        deploy_docker(config)
    elif config.deployment.method == 'podman':
        deploy_docker(config, use_podman=True)
    elif config.deployment.method == 'kubernetes':
        deploy_kubernetes(config)
    elif config.deployment.method == 'ansible':
        deploy_ansible(config)
    else:
        raise ValueError(f"Unsupported deployment method: {config.deployment.method}")

    print(f"CSV connectors deployed with log directory {config.csv_connectors.log_dir}")

def load_config(config_file):
    with open(config_file, 'r') as f:
        config_data = yaml.safe_load(f)
    return Config(**config_data)

if __name__ == "__main__":
    main()

File: containers/elastic/elastic__connectors/csv/rust/deploy_csv_connectors.rs
use std::fs::{self, File};
use std::io::Write;
use std::path::Path;
use std::process::Command;

use clap::{Arg, Command as ClapCommand};
use serde::{Deserialize, Serialize};
use serde_yaml;

#[derive(Serialize, Deserialize)]
struct CSVConnectorsConfig {
    log_dir: String,
    schemas_file: String,
}

#[derive(Serialize, Deserialize)]
struct SplunkConfig {
    enabled: bool,
    hec_url: String,
    hec_token: String,
}

#[derive(Serialize, Deserialize)]
struct ElasticsearchConfig {
    enabled: bool,
    host: String,
    index: String,
}

#[derive(Serialize, Deserialize)]
struct KubernetesConfig {
    replicas: i32,
    storage_class: String,
    log_storage_size: String,
}

#[derive(Serialize, Deserialize)]
struct DeploymentConfig {
    method: String,
    namespace: String,
    kubernetes: KubernetesConfig,
}

#[derive(Serialize, Deserialize)]
struct Config {
    csv_connectors: CSVConnectorsConfig,
    splunk: SplunkConfig,
    elasticsearch: ElasticsearchConfig,
    deployment: DeploymentConfig,
    buffer_timeout: f64,
    worker_count: i32,
    batch_size: i32,
    delimiter: String,
}

fn validate_resources() {
    let cpu_count = num_cpus::get();
    let mem_total = psutil::memory::virtual_memory().unwrap().total / 1024 / 1024 / 1024; // GB
    let disk_free = psutil::disk::disk_usage("/").unwrap().free / 1024 / 1024 / 1024; // GB
    if cpu_count < 2 {
        eprintln!("Warning: Less than 2 CPU cores available");
    }
    if mem_total < 2 {
        eprintln!("Warning: Less than 2GB memory available");
    }
    if disk_free < 2 {
        eprintln!("Warning: Less than 2GB disk space available");
    }
}

fn backup_files(config_files: &[&str], backup_dir: &str) {
    fs::create_dir_all(backup_dir).unwrap();
    for src in config_files {
        if Path::new(src).exists() {
            let dst = Path::new(backup_dir).join(Path::new(src).file_name().unwrap());
            fs::copy(src, dst).unwrap();
        }
    }
}

fn deploy_docker(config: &Config, use_podman: bool) -> Result<(), Box<dyn std::error::Error>> {
    let compose_cmd = if use_podman { "podman-compose" } else { "docker-compose" };
    if Command::new(compose_cmd).output().is_err() {
        return Err(format!("{} not found", compose_cmd).into());
    }

    Command::new(compose_cmd)
        .args(&["up", "-d", "--build"])
        .env("CSV_LOG_DIR", &config.csv_connectors.log_dir)
        .env("SCHEMAS_FILE", &config.csv_connectors.schemas_file)
        .env("SPLUNK_ENABLED", config.splunk.enabled.to_string())
        .env("SPLUNK_HEC_URL", &config.splunk.hec_url)
        .env("SPLUNK_TOKEN", &config.splunk.hec_token)
        .env("ES_ENABLED", config.elasticsearch.enabled.to_string())
        .env("ES_HOST", &config.elasticsearch.host)
        .env("ES_INDEX", &config.elasticsearch.index)
        .env("BATCH_SIZE", config.batch_size.to_string())
        .env("BUFFER_TIMEOUT", config.buffer_timeout.to_string())
        .env("WORKER_COUNT", config.worker_count.to_string())
        .env("CSV_DELIMITER", &config.delimiter)
        .status()?;

    Ok(())
}

fn deploy_kubernetes(config: &Config) -> Result<(), Box<dyn std::error::Error>> {
    if Command::new("kubectl").output().is_err() {
        return Err("kubectl not found".into());
    }

    let manifests = [
        "csv-pvc.yaml",
        "csv-splunk-deployment.yaml",
        "csv-elasticsearch-deployment.yaml",
    ];

    for manifest in &manifests {
        let content = fs::read_to_string(manifest)?;
        let content = content
            .replace("{{namespace}}", &config.deployment.namespace)
            .replace("{{replicas}}", &config.deployment.kubernetes.replicas.to_string())
            .replace("{{storage_class}}", &config.deployment.kubernetes.storage_class)
            .replace("{{log_storage_size}}", &config.deployment.kubernetes.log_storage_size)
            .replace("{{splunk_hec_url}}", &config.splunk.hec_url)
            .replace("{{splunk_token}}", &config.splunk.hec_token)
            .replace("{{es_host}}", &config.elasticsearch.host)
            .replace("{{es_index}}", &config.elasticsearch.index)
            .replace("{{batch_size}}", &config.batch_size.to_string())
            .replace("{{buffer_timeout}}", &config.buffer_timeout.to_string())
            .replace("{{worker_count}}", &config.worker_count.to_string())
            .replace("{{delimiter}}", &config.delimiter);

        let temp_file = format!("temp_{}", manifest);
        File::create(&temp_file)?.write_all(content.as_bytes())?;

        Command::new("kubectl")
            .args(&["apply", "-f", &temp_file])
            .status()?;

        fs::remove_file(&temp_file)?;
    }

    Ok(())
}

fn deploy_ansible(config: &Config) -> Result<(), Box<dyn std::error::Error>> {
    if Command::new("ansible-playbook").output().is_err() {
        return Err("ansible-playbook not found".into());
    }

    Command::new("ansible-playbook")
        .args(&["deploy_csv_connectors.yml", "-e", &format!("config_file={}", "deploy_config.yaml")])
        .status()?;

    Ok(())
}

fn cleanup_docker(use_podman: bool) -> Result<(), Box<dyn std::error::Error>> {
    let compose_cmd = if use_podman { "podman-compose" } else { "docker-compose" };
    if Command::new(compose_cmd).output().is_ok() {
        Command::new(compose_cmd)
            .args(&["down", "-v"])
            .status()?;
    }
    Ok(())
}

fn cleanup_kubernetes(config: &Config) -> Result<(), Box<dyn std::error::Error>> {
    if Command::new("kubectl").output().is_ok() {
        let manifests = [
            "csv-pvc.yaml",
            "csv-splunk-deployment.yaml",
            "csv-elasticsearch-deployment.yaml",
        ];
        for manifest in &manifests {
            Command::new("kubectl")
                .args(&["delete", "-f", manifest, "-n", &config.deployment.namespace])
                .status()?;
        }
    }
    Ok(())
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let matches = ClapCommand::new("Deploy CSV Connectors")
        .arg(Arg::new("config").default_value("deploy_config.yaml").help("Path to configuration file"))
        .arg(Arg::new("cleanup").long("cleanup").action(clap::ArgAction::SetTrue).help("Cleanup deployment"))
        .get_matches();

    let config_file = matches.get_one::<String>("config").unwrap();
    let cleanup = matches.get_flag("cleanup");

    let config_files = [
        "deploy_config.yaml",
        "docker-compose.yml",
        "Dockerfile",
        "schemas.yaml",
        "csv-splunk-deployment.yaml",
        "csv-elasticsearch-deployment.yaml",
        "csv-pvc.yaml",
        "deploy_csv_connectors.yml",
    ];

    if cleanup {
        println!("Cleaning up CSV connectors deployment...");
        let config = load_config(config_file)?;
        match config.deployment.method.as_str() {
            "docker" => cleanup_docker(false)?,
            "podman" => cleanup_docker(true)?,
            "kubernetes" => cleanup_kubernetes(&config)?,
            "ansible" => deploy_ansible(&config)?,
            _ => return Err(format!("Unsupported deployment method: {}", config.deployment.method).into()),
        }
        return Ok(());
    }

    validate_resources();
    let config = load_config(config_file)?;
    let backup_dir = format!("backup/{}", chrono::Local::now().format("%Y%m%d_%H%M%S"));
    backup_files(&config_files, &backup_dir);

    match config.deployment.method.as_str() {
        "docker" => deploy_docker(&config, false)?,
        "podman" => deploy_docker(&config, true)?,
        "kubernetes" => deploy_kubernetes(&config)?,
        "ansible" => deploy_ansible(&config)?,
        _ => return Err(format!("Unsupported deployment method: {}", config.deployment.method).into()),
    }

    println!("CSV connectors deployed with log directory {}", config.csv_connectors.log_dir);
    Ok(())
}

fn load_config(config_file: &str) -> Result<Config, Box<dyn std::error::Error>> {
    let file = File::open(config_file)?;
    let config: Config = serde_yaml::from_reader(file)?;
    Ok(config)
}